### ISSUE TITLE
Marshal.GetHRForException isn't setting up IErrorInfo on NativeAOT

### ISSUE THREAD

**Issue by Sergio0694** [dotnet/runtime#94088]
### Description

On NativeAOT, the `IErrorInfo` setup from `Marshal.GetHRForException` is not implemented:

https://github.com/dotnet/runtime/blob/568ed4125d0ae62c0d82d8f3703854fcc04bf32d/src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/PInvokeMarshal.cs#L29-L38

This causes a functional difference with CoreCLR, which instead correctly sets up all the `IErrorInfo` data:

https://github.com/dotnet/runtime/blob/568ed4125d0ae62c0d82d8f3703854fcc04bf32d/src/coreclr/vm/marshalnative.cpp#L594-L612

Notably, this seems to be guarded by `FEATURE_COMINTEROP`, so it seems it's not implemented when built-in COM interop is disabled, which is always the case on NativeAOT. That said, this feature on its own doesn't really _require_ built-in COM interop. We could probably take either approach here:
- Implement `IErrorInfo` on NAOT too (just with some blittable bindings and a P/Invoke, it's doable)
- Decide not to support this on NAOT, and just remove that `// TODO` and simplify that code

Having built-in support for `IErrorInfo` would certainly be nice.

Additional, if we did decide to remove it, one question I have is: suppose someone is writing a library that uses COM. They want to populate `IErrorInfo` for exceptions they catch in their native exports (eg. if they're building a vtable for a native object). The library could end up running under either CoreCLR (with or without built-in COM), or NAOT. How can they know whether the explicit `IErrorInfo` is needed on their end? And if they were to just implement this on their own and always do this, wouldn't this just be quite innefficient and cause extra unnecessary work when running on CoreCLR, as you'd be setting two `IErrorInfo` every single time (one from `Marshal.GetHRForException`, one from your own library code), and just discarding one?

### Reproduction Steps

n/a

### Expected behavior

n/a

### Actual behavior

n/a

**Comment by jkotas** [comment]
> Implement IErrorInfo on NAOT too (just with some blittable bindings and a P/Invoke, it's doable)

This method has a hard dependency on built-in COM interop. I do not think it is possible to create equivalent implementation with built-in COM interop.

> suppose someone is writing a library that uses COM. 

If the library wants to avoid dependency on built-in COM, I think it should avoid calling `Marshal.GetHRForException`.

**Comment by AaronRobinsonMSFT** [comment]
> That said, this feature on its own doesn't really require built-in COM interop. 

Yes it does. The conversion between `IErrorInfo` and .NET Exceptions are in play here.

> > suppose someone is writing a library that uses COM.

> If the library wants to avoid dependency on built-in COM, I think it should avoid calling Marshal.GetHRForException.

Agree.

Most of the `HRESULT` to/from `Exception` APIs on `Marshal` are rarely worth the trouble they tend to cause, especially when `IErrorInfo` is involved. If a user has an `Exception` and they want the `IErrorInfo` conversion done ideally it should be done at the marshaller layer. Right now we have a note in one of our marshallers, `ExceptionAsDefaultMarshaller<T>`, about depending on this but as we move forward to support more scenarios, that will change and we will raise that work to be done in the marshaller itself.

https://github.com/dotnet/runtime/blob/76a995afe3306863cb836b5becc33293a2e5a781/src/libraries/System.Runtime.InteropServices/src/System/Runtime/InteropServices/Marshalling/ExceptionAsDefaultMarshaller.cs#L6-L26

If someone wanted to update that marshaller to handle calling `SetLastError()` that would be most welcome.

**Comment by AaronRobinsonMSFT** [comment]
@jkoritzinsky and @jtschuster We should repurpose this issue to make `ExceptionAsDefaultMarshaller<T>`, or a new marshaller, do the calls to marshal the `Exception` to `IErrorInfo` as appropriate and call `SetLastError()`.
